database/sql/sql.go

```diff
@@ -190,6 +195,163 @@ func (s *sqlDatabase) cascadeMigration() error {
+		return nil
+	}
+
+	slog.Info("importing credentials from config")
+	for _, cred := range s.cfg.MigrateCredentials {
+		slog.Info("importing credential", "name", cred.Name)
+		parsed, err := url.Parse(cred.BaseEndpoint())
+		if err != nil {
+			return errors.Wrap(err, "parsing base URL")
+		}
+
+		certBundle, err := cred.CACertBundle()
+		if err != nil {
+			return errors.Wrap(err, "getting CA cert bundle")
+		}
+		hostname := parsed.Hostname()
+		createParams := params.CreateGithubEndpointParams{
+			Name:          hostname,
+			Description:   fmt.Sprintf("Endpoint for %s", hostname),
+			APIBaseURL:    cred.APIEndpoint(),
+			BaseURL:       cred.BaseEndpoint(),
+			UploadBaseURL: cred.UploadEndpoint(),
+			CACertBundle:  certBundle,
+		}
+
+		var endpoint params.GithubEndpoint
+		endpoint, err = s.GetGithubEndpoint(adminCtx, hostname)
+		if err != nil {
+			if !errors.Is(err, runnerErrors.ErrNotFound) {
+				return errors.Wrap(err, "getting github endpoint")
+			}
+			endpoint, err = s.CreateGithubEndpoint(adminCtx, createParams)
+			if err != nil {
+				return errors.Wrap(err, "creating default github endpoint")
+			}
+		}
+
+		credParams := params.CreateGithubCredentialsParams{
+			Name:        cred.Name,
+			Description: cred.Description,
+			Endpoint:    endpoint.Name,
+			AuthType:    params.GithubAuthType(cred.GetAuthType()),
+		}
+		switch credParams.AuthType {
+		case params.GithubAuthTypeApp:
+			keyBytes, err := cred.App.PrivateKeyBytes()
+			if err != nil {
+				return errors.Wrap(err, "getting private key bytes")
+			}
+			credParams.App = params.GithubApp{
+				AppID:           cred.App.AppID,
+				InstallationID:  cred.App.InstallationID,
+				PrivateKeyBytes: keyBytes,
+			}
+
+			if err := credParams.App.Validate(); err != nil {
+				return errors.Wrap(err, "validating app credentials")
+			}
+		case params.GithubAuthTypePAT:
+			token := cred.PAT.OAuth2Token
+			if token == "" {
+				token = cred.OAuth2Token
+			}
+			if token == "" {
+				return errors.New("missing OAuth2 token")
+			}
+			credParams.PAT = params.GithubPAT{
+				OAuth2Token: token,
+			}
+		}
+
+		creds, err := s.CreateGithubCredentials(adminCtx, credParams)
+		if err != nil {
+			return errors.Wrap(err, "creating github credentials")
+		}
+
+		if err := s.conn.Exec("update repositories set credentials_id = ?,endpoint_name = ? where credentials_name = ?", creds.ID, creds.Endpoint, creds.Name).Error; err != nil {
+			return errors.Wrap(err, "updating repositories")
+		}
+
+		if err := s.conn.Exec("update organizations set credentials_id = ?,endpoint_name = ? where credentials_name = ?", creds.ID, creds.Endpoint, creds.Name).Error; err != nil {
+			return errors.Wrap(err, "updating organizations")
+		}
+
+		if err := s.conn.Exec("update enterprises set credentials_id = ?,endpoint_name = ? where credentials_name = ?", creds.ID, creds.Endpoint, creds.Name).Error; err != nil {
+			return errors.Wrap(err, "updating enterprises")
+		}
```

Comment:
```suggestion
		creds, err := s.CreateGithubCredentials(adminCtx, credParams)
		if err != nil {
			return errors.Wrap(err, "creating github credentials")
		}

		if err := s.conn.Exec("update repositories set credentials_id = ?,endpoint_name = ? where credentials_name = ?", creds.ID, creds.Endpoint.Name, creds.Name).Error; err != nil {
			return errors.Wrap(err, "updating repositories")
		}

		if err := s.conn.Exec("update organizations set credentials_id = ?,endpoint_name = ? where credentials_name = ?", creds.ID, creds.Endpoint.Name, creds.Name).Error; err != nil {
			return errors.Wrap(err, "updating organizations")
		}

		if err := s.conn.Exec("update enterprises set credentials_id = ?,endpoint_name = ? where credentials_name = ?", creds.ID, creds.Endpoint.Name, creds.Name).Error; err != nil {
			return errors.Wrap(err, "updating enterprises")
		}
```

`creds.Endpoint` is of type `GithubEndpoint` ([xref](https://github.com/gabriel-samfira/garm/blob/8726cb994eeeee709a98ae49f2d793f35646d6a1/params/params.go#L825-L834)) and can be converted.

in my test, garm-server crashed with the following output:

```
sql: converting argument $2 type: unsupported type params.GithubEndpoint, a struct\nupdating repositories\ngithub.com/cloudbase/garm/database/sql.(*sqlDatabase).migrateCredentialsToDB\n\t/go/garm_repo/database/sql/sql.go:341\ngithub.com/cloudbase/garm/database/sql.(*sqlDatabase).migrateDB\n\t/go/garm_repo/database/sql/sql.go:428\ngithub.com/cloudbase/garm/database/sql.NewSQLDatabase\n\t/go/garm_repo/database/sql/sql.go:77\ngithub.com/cloudbase/garm/database.NewDatabase\n\t/go/garm_repo/database/database.go:30\nmain.main\n\t/go/garm_repo/cmd/garm/main.go:175\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:267\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650"
time=2024-05-07T09:57:43.562Z level=ERROR msg="rolling back github github endpoints table" error="sql: converting argument $2 type: unsupported type params.GithubEndpoint, a struct\nupdating repositories\ngithub.com/cloudbase/garm/database/sql.(*sqlDatabase).migrateCredentialsToDB\n\t/go/garm_repo/database/sql/sql.go:341\ngithub.com/cloudbase/garm/database/sql.(*sqlDatabase).migrateDB\n\t/go/garm_repo/database/sql/sql.go:428\ngithub.com/cloudbase/garm/database/sql.NewSQLDatabase\n\t/go/garm_repo/database/sql/sql.go:77\ngithub.com/cloudbase/garm/database.NewDatabase\n\t/go/garm_repo/database/database.go:30\nmain.main\n\t/go/garm_repo/cmd/garm/main.go:175\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:267\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650"
time=2024-05-07T09:57:43.562Z level=INFO msg="migrating database: migrating credentials: updating repositories: sql: converting argument $2 type: unsupported type params.GithubEndpoint, a struct"
```

Reply:
Added a DB migration test.